Implement config validation

iojw opened the pull request:
<!-- Describe the changes in this PR -->

Closes #2474 

Implements validation for the `~/.sky/config.yaml` file. If there is an error, an exception is thrown and the user is warned.

To avoid circular imports, I had to refactor the code a bit, moving validation code into `common_utils` and moving the imports in `sky.util.schemas` to within the functions.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Pass existing config tests, added new unit tests
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

Michaelvll reviewed (commented):
Thanks for submitting the PR @iojw! The code looks mostly good to me, but there seems to be a circular import, when I tried `sky check` with the current PR.
```
sky/data/storage.py", line 49, in <module>
    str(clouds.AWS()),
AttributeError: partially initialized module 'sky.clouds' has no attribute 'AWS' (most likely due to a circular import)
```

iojw commented on sky/skypilot_config.py:
```diff
@@ -160,6 +161,15 @@ def _try_load_config() -> None:
             logger.debug(f'Config loaded:\n{pprint.pformat(_dict)}')
         except yaml.YAMLError as e:
             logger.error(f'Error in loading config file ({config_path}):', e)
+        try:
+            if _dict is not None:
+                common_utils.validate_schema(_dict,
+                                             schemas.get_config_schema(),
+                                             'Invalid config YAML: ',
+                                             skip_none=False)
+        except ValueError as e:
+            logger.error(f'Error loading config file ({config_path}: {e}')
+            _dict = None
```
Yes, I was deciding between these 2 options as well. What do you think? I was thinking that a log is visible enough, but I'm also OK with raising an exception.

Michaelvll replied:
cc'ing @concretevitamin as the owner of this code : )

concretevitamin replied:
Yeah it’s not a clear one. I lean towards raising. We raise in _syntax_check_for_ssh_proxy_command() anyway, and raising will help catch issues such as incorrect indentation.

iojw commented:
Updated to throw an exception and also to fix the circular import issue mentioned (it's caused by the `kubernetes_utils` import in `schemas.py` so I've moved the enums to a separate module).

Strangely, I'm unable to reproduce it locally, so could you help to verify that it is fixed @Michaelvll ?

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -260,3 +261,126 @@ def get_cluster_schema():
+                'required': [],
+                'additionalProperties': False,
+                'properties': {
+                    'specific_reservations': {
+                        'type': 'array',
+                        'items': {
+                            'type': 'string',
+                            'minItems': 1,
+                            'maxItems': 1,
+                        }
+                    },
```
Not sure why, but I can still specify more than 1 `specific_reservations`

iojw replied:
Yes there's a bug in my schema, thank you for catching this! Also added a test to confirm this.

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -4,11 +4,11 @@
 def get_resources_schema():
+    # To avoid circular imports, only import when needed.
+    # pylint: disable=import-outside-toplevel
+    from sky.clouds.service_catalog import _ALL_CLOUDS
```
style nit: import modules instead of object. 

Also, seems kubernetes is not included in `_ALL_CLOUDS`. Is it possible to still make the `cloud_registry` work?

iojw replied:
I think relying on `cloud_registry` for the list of clouds for validating `skypilot_config` will be very brittle because of the many potential circular imports. This is because for `cloud_registry` to be updated with all the clouds, we need all the cloud modules in `sky.clouds.aws/gcp/...` to be imported. However, there are many modules that require `skypilot_config` before this is done.

For example, the following import path: `sky.clouds` -> `sky.clouds.cloud` -> `sky.skypilot_config` occurs before any cloud module `sky.clouds.aws/gcp/...` is imported, and changing this order will lead to a circular import, as seen from the commment in `sky.clouds.__init__.py`. Currently, I don't see how we can avoid this.

To include kubernetes in `_ALL_CLOUDS`, I believe we can do a slight adjustment to the logic in `sky.clouds.service_catalog.__init__.py` instead. Please take a look!

Michaelvll reviewed (commented):
Thanks for identifying the issues and fixing them @iojw! It looks mostly good. Left some minor comments. cc'ing @concretevitamin to see if he has some additional test to be run. : )

iojw commented on sky/utils/kubernetes_enums.py:
```diff
@@ -0,0 +1,29 @@
+"""Kubernetes enums for SkyPilot."""
```
Yes, we need to have the enums outside of `kubernetes_utils` because this will lead to many circular imports. e.g.
`sky.skypilot_config -> sky.utils.common_utils -> sky.utils.schemas -> sky.utils.kubernetes_untils -> sky.utils.backend_utils -> sky.skypilot_config`

iojw commented on sky/utils/schemas.py:
```diff
@@ -260,3 +261,126 @@ def get_cluster_schema():
+                    },
+                }
+            },
+            'kubernetes': {
+                'type': 'object',
+                'required': [],
+                'additionalProperties': False,
+                'properties': {
+                    'networking': {
+                        'type': 'string',
+                        'case_insensitive_enum': [
+                            type.value for type in
+                            kubernetes_enums.KubernetesNetworkingMode
+                        ]
+                    },
+                }
+            },
```
Yup I will add this! In the future, we should remember to have the doc updated if any new PRs add new config fields.

@romilbhardwaj Could you confirm that the docs I've written are correct?

iojw commented on sky/utils/schemas.py:
```diff
@@ -260,3 +261,126 @@ def get_cluster_schema():
+                    },
+                }
+            },
+            'oci': {
+                'type': 'object',
+                'required': [],
+                # Properties are either 'default' or a region name.
+                'additionalProperties': {
+                    'type': 'object',
+                    'required': [],
+                    'additionalProperties': False,
+                    'properties': {
+                        'compartment_ocid': {
+                            'type': 'string',
+                        },
+                        'image_tag_general': {
+                            'type': 'string',
+                        },
+                        'image_tag_gpu': {
+                            'type': 'string',
+                        },
+                        'vcn_subnet': {
+                            'type': 'string',
+                        },
+                    }
+                },
+            },
+        }
```
Updated, please take a look at the docs as I'm not very familiar with OCI (cc @HysunHe as well!)

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -4,11 +4,11 @@
+    # pylint: disable=import-outside-toplevel
+    from sky.clouds import service_catalog
     return {
         '$schema': 'http://json-schema.org/draft-07/schema#',
```
For code gardening: why do we use a different schema for resources than the others?

iojw replied:
Hm, I'm not sure actually. I've updated all of the schemas to the newest one to be consistent, and tested that it still works as expected.

Michaelvll reviewed (approved):
Thanks for adding the schema check @iojw! LGTM with an additional comment. Would be nice to do some more manual test to avoid any accident false positive schema alert.

iojw commented on docs/source/reference/config.rst:
```diff
@@ -113,3 +113,32 @@ Available fields and semantics:
+    #
+    # nodeport: Exposes a service on a static port number on each Node, allowing external access to services using <NodeIP>:<NodePort>.
+    # portforward: Forward a local port to a port on a Pod, typically used for debugging or temporary access purposes.
+    networking:  nodeport
```
Done!

romilbhardwaj commented on docs/source/reference/config.rst:
```diff
@@ -113,3 +113,33 @@ Available fields and semantics:
+    #
+    # nodeport: Exposes a service on a static port number on each Node, allowing external access to services using <NodeIP>:<NodePort>.
+    # portforward: Forward a local port to a port on a Pod, typically used for debugging or temporary access purposes.
+    networking:  nodeport
```
There seem to be two spaces here :) Also, should we default to portforward here?
```suggestion
    networking: portforward
```

romilbhardwaj commented on docs/source/reference/config.rst:
```diff
@@ -113,3 +113,33 @@ Available fields and semantics:
+
+  # Advanced Kubernetes configurations (optional).
+  kubernetes:
+    # The networking mode for accessing jump pods (optional).
```
```suggestion
    # The networking mode for accessing SSH jump pod (optional).
```

romilbhardwaj commented on docs/source/reference/config.rst:
```diff
@@ -113,3 +113,33 @@ Available fields and semantics:
+  # Advanced Kubernetes configurations (optional).
+  kubernetes:
+    # The networking mode for accessing jump pods (optional).
+    # This must be either: 'nodeport' or 'portforward'.
```
```suggestion
    # This must be either: 'nodeport' or 'portforward'. If not specified, defaults to 'portforward'.
```

romilbhardwaj commented on docs/source/reference/config.rst:
```diff
@@ -113,3 +113,33 @@ Available fields and semantics:
+    # The networking mode for accessing jump pods (optional).
+    # This must be either: 'nodeport' or 'portforward'.
+    #
+    # nodeport: Exposes a service on a static port number on each Node, allowing external access to services using <NodeIP>:<NodePort>.
```
```suggestion
    # nodeport: Exposes the jump pod SSH service on a static port number on each Node, allowing external access to using <NodeIP>:<NodePort>. Using this mode requires opening multiple ports on nodes in the Kubernetes cluster.
```

romilbhardwaj commented on docs/source/reference/config.rst:
```diff
@@ -113,3 +113,33 @@ Available fields and semantics:
+    # This must be either: 'nodeport' or 'portforward'.
+    #
+    # nodeport: Exposes a service on a static port number on each Node, allowing external access to services using <NodeIP>:<NodePort>.
+    # portforward: Forward a local port to a port on a Pod, typically used for debugging or temporary access purposes.
```
```suggestion
    # portforward: Uses `kubectl port-forward` to create a tunnel and directly access the jump pod SSH service in the Kubernetes cluster. Does not require opening ports the cluster nodes and is more secure. 'portforward' is used as default if 'networking' is not specified.
```

romilbhardwaj reviewed (commented):
Thanks @iojw! Left some comments on the k8s config docs, otherwise LGTM!

iojw commented:
Thank you @romilbhardwaj for taking a look! I'll merge this in soon if there are no further comments.
